### ISSUE TITLE
Typeguard for assignment expression

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#1116]
```python
from typing import Union

def foo(x: Union[str, int]):
    if isinstance((y := x), str):
        z: str = y  # Error, though y is `str`
    if isinstance(x, str):
        z: str = x  # Ok
```

Error:

```
  5:18 - error: Expression of type "str | int" cannot be assigned to declared type "str"
    Type "str | int" cannot be assigned to type "str"
      "int" is incompatible with "str" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos 
```

pyright version: 6.14.8

Typeguard doesn't seem applicable to the Assignment Expression. Is it a bug or intended?
If intended, I'll submit this as Feature Request.

**Comment by tamuhey** [comment]
mypy doesn't complain to the code.

**Comment by erictraut** [comment]
I don't think this pattern is common enough to justify adding special conditional type narrowing logic. I'll leave the issue open for a little while to see if others upvote it or comment.

**Comment by erictraut** [comment]
Why would you use this pattern in the first place? As you note, type narrowing does apply in the case where you don't use a type assignment expression within the `isinstance` call. What is the value in assigning the reference to another local variable?

**Comment by tamuhey** [comment]
Roughly speaking, I was trying to run the following code:

```python
def foo(cfg: Dict[str, Union[str, Dict[str]]]) -> str:
    if isinstance((v := cfg["foo"]), str):
        return v # Error
    ...
```

The following raises error:

```python
def foo(cfg: Dict[str, Union[str, Dict[str]]]) -> str:
    v = cfg["foo"]
    if isinstance(v, str):
        return cfg["foo"]
    ...
```

So I tried the following next:

```python
def foo(cfg: Dict[str, Union[str, Dict[str]]]) -> str:
    v = cfg["foo"]
    if isinstance(v, str):
        return v
    ...
```

This is where assignment expression and typeguard come in.

**Comment by erictraut** [comment]
I see, thanks for the additional details.

Yeah, as you noted, there's a simple workaround — one that arguably improves the readability of the code.

**Comment by erictraut** [comment]
I think this is too esoteric to justify adding the additional logic, so I'm going to close the request. If other users want to upvote it or add comments, we might reconsider this decision in the future.
